packages/daemons/blocks_collection.go

```diff
@@ -184,64 +186,161 @@ func UpdateChain(ctx context.Context, d *daemon, host string, maxBlockID int64)
 		return err
 	}
 
-	for blockID := curBlock.BlockID + 1; blockID <= maxBlockID; blockID++ {
-		if ctx.Err() != nil {
-			d.logger.WithFields(log.Fields{"type": consts.ContextError, "error": ctx.Err()}).Error("context error")
-			return ctx.Err()
-		}
+	if ctx.Err() != nil {
+		d.logger.WithFields(log.Fields{"type": consts.ContextError, "error": ctx.Err()}).Error("context error")
+		return ctx.Err()
+	}
+
+	type rbr struct {
+		blockID  int64
+		rawBlock []byte
+	}
 
+	// getRawBlock is 1st conveyor stage. It extracts the block body through a semaphore and sending raw body to 2st stage
+	getRawBlock := func(rawBlocksCh chan rbr, semaphore chan bool, errCh chan error, host string, blockID int64) {
+		<-semaphore
 		blockBin, err := utils.GetBlockBody(host, blockID, consts.DATA_TYPE_BLOCK_BODY)
```

Comment:
That is the most funny part, yes CONCURRENT CONVEYOUR is cool, but did you read the task?
Because the main bottleneck there is that utils.GetBlockBody function, because it connects to other node on every fucking block, may it would be better to download 1000 blocks at one time???